r/tests/testthat/test-parquet.R

```diff
@@ -484,3 +484,43 @@ test_that("Can read Parquet files from a URL", {
   expect_true(tibble::is_tibble(pu))
   expect_identical(dim(pu), c(10L, 11L))
 })
+
+test_that("thrift string and container size can be specified when reading Parquet files", {
+
+  tf <- tempfile()
+  on.exit(unlink(tf))
+  table <- arrow_table(example_data)
+  write_parquet(table, tf)
+  file <- make_readable_file(tf)
+  on.exit(file$close())
+
+  # thrift string size
+  reader_props <- ParquetReaderProperties$create()
+  reader_props$set_thrift_string_size_limit(1)
+  expect_identical(reader_props$thrift_string_size_limit(), 1L)
+
+  # We get an error if we set the Thrift string size limit too small
+  expect_error(ParquetFileReader$create(file, reader_props = reader_props), "TProtocolException: Exceeded size limit")
+
+  # Increase the size and we can read successfully
+  reader_props$set_thrift_string_size_limit(10000)
+  reader <- ParquetFileReader$create(file, reader_props = reader_props)
+  data <- reader$ReadTable()
+  expect_identical(collect.ArrowTabular(data), example_data)
```

Comment:
Given this isn't urgent to merge, I'd like to slow down here and think at a more general level about what we are testing.  Part of the problem here which has forced us into testing C++ messages is the fact that we're testing things we don't control.  

I started thinking we should only test that we can get/set these values, but not test the outcomes of running the code.  However, this was because I'd assumed that setting `thrift_string_size_limit` would be tested in the C++ layer, but after looking at the original PR, #13275, I see that it is not, and is testing in PyArrow, so we should probably follow that pattern and consider both the unit tests (i.e. are our R objects working as expected) and integration tests (i.e. are the returned values as expected) our responsibility.  

I did have thoughts about splitting out tests into separate files for integration tests versus unit tests, but I think this would be more of an exercise in philosophical pedantry than adding actual value or making our lives easier.

Reply:
> Testing the C++ message is less about testing C++ behaviour and 100% about making sure we are testing the code path that we think we are testing

I don't get this, can you expand? I was thinking that as we're testing the outcome of setting the thrift string value ridiculously low and the fact that it does cause an error, then that's testing the C++ behaviour, and so is more integration testing (but honestly, probably so are most of our tests then).  We can also take this chat offline for our catchup if you like, as while it's interesting, it's not necessarily all that relevant to us merging this PR ;)